toolchain/check/impl.cpp

```diff
@@ -227,17 +232,197 @@ auto ImplWitnessForDeclaration(Context& context, const SemIR::Impl& impl)
+  auto facet_type = context.types().GetAs<SemIR::FacetType>(type_id);
+  auto facet_info = context.facet_types().Get(facet_type.facet_type_id);
+  if (auto impls = facet_info.TryAsSingleInterface()) {
+    return impls->interface_id == interface_id;
+  }
+  return false;
+}
+
+auto AddConstantsToImplWitnessFromConstraint(Context& context,
+                                             const SemIR::Impl& impl,
+                                             SemIR::InstId witness_id,
+                                             SemIR::ImplId prev_decl_id)
     -> void {
+  CARBON_CHECK(!impl.has_definition_started());
+  CARBON_CHECK(witness_id.is_valid());
+  if (witness_id == SemIR::ErrorInst::SingletonInstId) {
+    return;
+  }
+  auto facet_type_id = context.GetTypeIdForTypeInst(impl.constraint_id);
+  CARBON_CHECK(facet_type_id != SemIR::ErrorInst::SingletonTypeId);
+  auto facet_type = context.types().GetAs<SemIR::FacetType>(facet_type_id);
+  const SemIR::FacetTypeInfo& facet_type_info =
+      context.facet_types().Get(facet_type.facet_type_id);
+
+  auto interface_type = facet_type_info.TryAsSingleInterface();
+  CARBON_CHECK(interface_type.has_value());
+  const auto& interface =
+      context.interfaces().Get(interface_type->interface_id);
+
+  auto witness = context.insts().GetAs<SemIR::ImplWitness>(witness_id);
+  auto witness_block = context.inst_blocks().GetMutable(witness.elements_id);
+  auto assoc_entities =
+      context.inst_blocks().Get(interface.associated_entities_id);
+  CARBON_CHECK(witness_block.size() == assoc_entities.size());
+
+  // Scan through rewrites, produce map from element index to constant value.
+  // TODO: Perhaps move this into facet type resolution?
+  llvm::SmallVector<SemIR::ConstantId> rewrite_values(
+      assoc_entities.size(), SemIR::ConstantId::Invalid);
+  for (auto rewrite : facet_type_info.rewrite_constraints) {
+    auto inst_id = context.constant_values().GetInstId(rewrite.lhs_const_id);
+    auto access = context.insts().GetAs<SemIR::ImplWitnessAccess>(inst_id);
+    if (!WitnessAccessMatchesInterface(context, access.witness_id,
+                                       interface_type->interface_id)) {
+      // Skip rewrite constraints that apply to associated constants of
+      // a different interface than the one being implemented.
+      continue;
+    }
+    CARBON_CHECK(access.index.index >= 0);
+    CARBON_CHECK(access.index.index <
+                 static_cast<int32_t>(rewrite_values.size()));
+    auto& rewrite_value = rewrite_values[access.index.index];
+    if (rewrite_value.is_valid()) {
+      if (rewrite_value != rewrite.rhs_const_id) {
+        // TODO: Do at least this checking as part of facet type resolution
+        // instead.
+
+        // TODO: Figure out how to print the two different values
+        // `rewrite_value` & `rewrite.rhs_const_id` in the diagnostic message.
+        CARBON_DIAGNOSTIC(AssociatedConstantWithDifferentValues, Error,
```

Comment:
We should skip producing the diagnostic (here and the similar checks below) if either of the values is `ErrorInst`.

Reply:
I stepped through this in the debugger. What is happening is that eval of the `where` expression gets `Phase::UnknownDueToError` if any of the constraints has an error. Which means the constant value of a `where` expression with an error is `ErrorInst`. So we get `impl ... as <Error>` and downstream code never even sees the rewrite constraints.